fix folder already rendered with same name

enricosada opened the pull request:
fix #246

latkin commented on vsintegration/src/unittests/Tests.ProjectSystem.MSBuildUtilities.fs:
```diff
@@ -0,0 +1,74 @@
+        |> allPathsSeparator
+        |> Seq.map (fun (a,b,result) -> TestCaseAttribute(a, b, result))
+
+    [<Test; TestCaseSource("IsSameDirTestCases")>]
```
Our internal build is still on nunit 2.2.2, so `TestCaseSource` is not available.

enricosada replied:
i dont understand, internal ci doesn't run the repo bat files for tests?

latkin replied:
Internally, source is in big TFS repo along with other Visual Studio stuff, and has various fixed sets of dependencies. Projects used to build shipping VF# tools are not the same as in this repo (though close), tests are run through different systems.  We are looking at unifying/simplifying, but for now NUnit stuff needs to be compatible with 2.2.2.

latkin commented on vsintegration/src/vs/FsPkgs/FSharp.Project/FS/MSBuildUtilities.fs:
```diff
@@ -12,6 +12,60 @@ open System.Diagnostics
+    let FolderComparer = StringComparer.OrdinalIgnoreCase 
+
+    let GetPathParts (x:string) =
+        x.Split( [| '\\' |], StringSplitOptions.RemoveEmptyEntries) |> Seq.ofArray
```
Why convert to `Seq`?

latkin commented on vsintegration/src/vs/FsPkgs/FSharp.Project/FS/MSBuildUtilities.fs:
```diff
@@ -12,6 +12,60 @@ open System.Diagnostics
+            match b, d with
+            | [], [] -> false
+            | [], _ -> true
+            | _, [] -> false
```
This case is not needed I don't think.

Could flatten this even more if you want:
```fsharp
let rec helper b d = 
    match b, d with
    | [], _ :: _ -> true
    | x :: xs, y :: ys when FolderComparer.Compare(x,y) = 0  -> helper xs ys
    | _ -> false
```

latkin commented on vsintegration/src/vs/FsPkgs/FSharp.Project/FS/MSBuildUtilities.fs:
```diff
@@ -12,6 +12,60 @@ open System.Diagnostics
+    let IsSameDir a b =
+        0 = Seq.compareWith (fun x y -> FolderComparer.Compare(x,y)) (GetPathParts a) (GetPathParts b)
+
+    let IsSubDirOf baseDir directory  =
```
`IsSubDirOf` and `IsSameDir` are always used together, with the same args, with `||`.  Does it make more sense to just create one function `isSameOrSubDir`?  Would be just small modification of this function (`[], []` case returns true).

enricosada replied:
i separated the two function because are easier to test
I lost unit tests on `IsSubDirOf` and `IsSameDir` on rebase i think :sweat_smile:

latkin replied:
If they are always used together, they can still be combined and tested as a single function.

latkin commented on vsintegration/src/vs/FsPkgs/FSharp.Project/FS/MSBuildUtilities.fs:
```diff
@@ -12,6 +12,60 @@ open System.Diagnostics
 open System.IO
 open System
 
+module internal MSBuildHelpers =
```
Maybe it's just me, but I am having a very hard time following the logic in here... Maybe some comments with examples and/or more descriptive variable names would help?

latkin commented on vsintegration/src/vs/FsPkgs/FSharp.Project/FS/MSBuildUtilities.fs:
```diff
@@ -108,6 +159,24 @@ type internal MSBuildUtilities() =
+                dir :: visitedDirs
+
+        EnumerateItems(big)
+        |> Seq.map (fun bi -> bi, GetUnescapedUnevaluatedInclude(bi) |> relativePath)
```
The `relativePath` function, and first line of `checkAlreadyRenderedFolder` are together doing the same thing as existing function `ComputeFolder`.

Recommend you just call that function here, and remove `relativePath` function and first line of `checkAlreadyRenderedFolder`

latkin commented on vsintegration/src/.gitignore:
```diff
@@ -0,0 +1 @@
+TestResults
```
TestResults folder should exist only under %root%\tests. Why ignore it here?

enricosada replied:
when i run unit test from inside visual studio, this is the folder created.
I'll double check

enricosada replied:
@latkin i tried, when i run test from visual studio test runner i have
    
    vsintegration/src/TestResults/Deployment.TE.Tests.mdf
    vsintegration/src/TestResults/Deployment.TE.Tests_log.ldf

is the wrong directory?

latkin replied:
I did not realize anyone was running the unit tests through VS test runner. Is there some config that can be added to the project so that logs go in the normal place under %root%\tests\TestResults?

latkin commented on vsintegration/src/unittests/Tests.ProjectSystem.MSBuildUtilities.fs:
```diff
@@ -0,0 +1,74 @@
+    member public this.IsSubDirOf baseDir dir expected = 
+        Assert.AreEqual(expected, dir |> MSBuildHelpers.IsSubDirOf baseDir)
+
+    static member IsAlreadyRenderedFolderTestCases =
```
Can you clarify the logic on these tests? Does left side list represent already-rendered folders, and right side is new folder being rendered -- and return val represents the shared subdirs between them?

```fsharp
[@"a\e\d"; @"a\b"], @"a\b", None                  // why not "a\b"?
[@"a\b"], @"b", None                              // ok
[@"a\b"; @"a\e\d"], @"a\b", Some (@"a\b")         // ok
[@"a\b\c"; @"b"], @"a", Some (@"a")               // ok
[@"a\b\c"; @"b"], @"a\b", Some (@"a")             // why not "a\b"?
[@"a\b\c"; @"b"], @"a\b\d", Some (@"a")           // why not "a\b"?
[@"a\b\c"; @"a"], @"a", None                      // why not "a"?
[@"a\b\c"; @"a"], @"a\b", Some (@"a\b")           // why not "a\b"?
[@"a\b\c"; @"a"], @"a\b\c", Some (@"a\b")         // why not "a\b\c"?
[@"a\b\c"; @"a"], @"a\b\e", Some (@"a\b")         // ok
[@"a\b\c\d"; @"a\b"], @"a\b\c\e", Some (@"a\b\c") // ok
```

enricosada replied:
yes pretty much, but shared is partial, because for example

```[@"a\b\c"; @"a"], @"a\b\c", Some (@"a\b") ```

`a\b` is the folder who is invalid, not `a\b\c`, because `a\b` is the first folder rendered twice

the logic is a bit annoing, i'll try to cleanup and document more

latkin replied:
In `a\b\c` is `c` a folder or a file?  I don't get how specifying the folder `a\b\c` twice means only `a\b` is shared?

enricosada replied:
they are all folders, with inside a file

because for rendere `a\b\c` you need to renderer

    A                                                                            
    +                                                                            
    +-+B                                                                         
    |  +                                                                         
    |  +-+C                                                                      
    |                                                                            
    +-+file.fs                                                                   
    |                                                                            
    +-+B                                                                         
       +                                                                         
       +-+C                                                                      

the first folder rendered two time is a\b

latkin commented:
Thanks for the PR, this will be great to have fixed.  Looks like a good set of tests.

Overall I'd just recommend adding some comments and refactorings so that it's more clear how exactly this gets validated, and what's considered valid/invalid.

enricosada commented:
thx for review @latkin , i'll send an update

enricosada commented:
Sry for delay. 
During refactoring i tried to fix root problem (add multiple nodes with same name) but is not ready yet.
I'll complete this one and update it, max this weekend, and send the other fix as separate pr

enricosada commented:
Thx for Ping :)

KevinRansom commented:
Enrico, is there any progress to report on this?

Thank you

Kevin

enricosada commented:
i am near to end, i used lot of time for complete but is going nice. 
i need to update status more while working

KevinRansom commented:
What you do is fine mate, and thank you so much for taking on this work.

Kevin

yupferris commented:
status? :)

latkin commented:
@enricosada I've heard before that this bug is annoying, it will be great to have your fix.  Do you have an ETA on an update?

enricosada commented:
i added the new version.
i'll check why build failed, but i rebased this branch onto master HEAD before push

the idea is to used a tree structure (the RoseTree) and check the tree for the invalid folder logic.
The function also remove unused msbuild `<Folder ..>` items

Every node is a ProjectTreeNode:
- Folder, the virtual folder used by solution explorer
- Item, an item with a reference to the msbuild item

every node can have children,  exactly like solution explorer, so there is not  a 1-1 mapping with msbuild items 
 
the function explode every msbuild item as tree, add every tree to the project root, and discard unused folders

the tree is checked for invalid folders

The algorithm use domain types, so we can unit test, instead of run the integration test with full visual studio (slow).
I think this is good because we can start take advantage of f# discriminated union for msbuild nodes (like ComReference etc) and behaviours

enricosada commented:
btw, can i add the [FsCheck](https://github.com/fsharp/FsCheck) as reference from nuget? 
For vs extension tests (not core, so np for profiles)

sometime is easier to use that for tests (like path, tree, list) because easy invariant

KevinRansom commented:
I like this, it is a big ambitious fix.  We will have to go over the code more thoroughly but it certainly looks, nice with a lot of detailed test cases, thank you for this.

However, given where we are in the cycle and how late it is, I think we are going to have to postpone taking this until the next version.  We are quite thoroughly locked down right now, and accepting changes only with the minimum necessary footprint changes.  It looks like a bunch of opportunistic and good restructuring was done, but we do not really have the runway to ensure that such a big set of changes to our project system do not have any impactful side effects.

It is a shame, I was hoping that we would ship Dev 14 with a fix for the issue.  We will not decline this pull just keep it until we can accept it into F# V.Next.  People should continue to review and thoroughly vet this submission.

msftclas commented:
__@enricosada__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

latkin commented:
@enricosada is this big change still beneficial/needed, given that the original bug was fixed?

latkin commented:
Closing this out as original bug has been fixed.  Please migrate this work to the tip of OOB and re-open if you'd like to revisit this for an upcoming release.
